src/hotblocks.rs

```diff
@@ -120,38 +120,105 @@ pub fn register_metrics(registry: &mut Registry, hotblocks: Arc<HotblocksHandle>
             source = %url,
-            "Starting periodic block processing time measurement task"
+            "Starting block processing time measurement for head updates"
         );
         
-        tokio::spawn(async move {
-            let mut interval = tokio::time::interval(std::time::Duration::from_secs(15));
-            
-            loop {
-                interval.tick().await;
+        match hotblocks.server.subscribe_head_updates(dataset_id) {
+            Ok(mut head_receiver) => {
+                tokio::spawn(async move {
+                    {
+                        let current_head = head_receiver.borrow().clone();
+                        if let Some(head) = current_head {
+                            let block_number = head.number;
+                            let block_hash = head.hash;
+                            
+                            tracing::debug!(
+                                dataset = %dataset_str,
+                                source = %url,
+                                block = %block_number,
+                                "Measuring block processing time for initial head"
+                            );
+                            
+                            crate::metrics::report_block_available(
+                                &client,
+                                &dataset_str,
+                                &network, 
+                                block_number,
+                                &block_hash,
+                            ).await;
+                        }
+                    }
+                    
+                    while head_receiver.changed().await.is_ok() {
+                        let current_head = head_receiver.borrow().clone();
+                        if let Some(head) = current_head {
+                            let block_number = head.number;
+                            let block_hash = head.hash;
+                            
+                            tracing::debug!(
+                                dataset = %dataset_str,
+                                source = %url,
+                                block = %block_number,
+                                "Measuring block processing time for new head"
+                            );
+                            
+                            crate::metrics::report_block_available(
+                                &client,
+                                &dataset_str,
+                                &network, 
+                                block_number,
+                                &block_hash,
+                            ).await;
+                        }
+                    }
+                    
+                    tracing::warn!(
+                        dataset = %dataset_str,
+                        source = %url,
+                        "Head update subscription ended"
+                    );
+                });
+            },
+            Err(err) => {
```

Comment:
This can only fail if the dataset was not found, so we can always predict it from the client side. If you're using the same datasets list that was used to build the HotblockServer, you can simply `unwrap`/`expect` this result